tests/test_TFNetworkLayer.py

```diff
@@ -3065,6 +3065,71 @@ def mask_func(source, **kwargs):
       last_var_v = var_v
 
 
+def test_reduce_mean_in_time():
+  with make_scope() as session:
+    import numpy as np
+    n_out = 5
+    config = Config({
+      "debug_print_layer_output_template": True,
+      "extern_data": {
+        "data": {"dim": n_out},
+        "classes": {"dim": n_out},
+      }})
+    net = TFNetwork(config=config, train_flag=True)
+    net.construct_from_dict({
+      "reduce_mean": {"class": "reduce", "mode": "mean", "axis": "T", "keep_dims": True, "from": ["data"],
+                      "is_output_layer": True}
+    })
+    session.run(tf.global_variables_initializer())
+    out = net.layers["reduce_mean"].output.placeholder
+    n_batch = 3
+    max_seq_len = 10
+    feed = make_feed_dict(net.extern_data.data.values(), n_batch=n_batch, n_time=max_seq_len, same_time=True)
+    v = session.run(out, feed_dict=feed)
+    input_len = feed[list(feed.keys())[0]]
+    input_data = feed[list(feed.keys())[1]]
+
+    ref = np.zeros([n_batch, n_out])
+    for batch, len in enumerate(input_len):
+      ref[batch, :] = np.mean(input_data[batch, :len, :], axis=0)
+    ref = np.expand_dims(ref, axis=1)
+
+    np.testing.assert_allclose(ref, v, rtol=1e-5)
+
+
+def test_reduce_mean_batch_time():
+  with make_scope() as session:
+    import numpy as np
+    n_out = 5
+    config = Config({
+      "debug_print_layer_output_template": True,
+      "extern_data": {
+        "data": {"dim": n_out},
+        "classes": {"dim": n_out},
+      }})
+    net = TFNetwork(config=config, train_flag=True)
+    net.construct_from_dict({
+      "reduce_mean": {"class": "reduce", "mode": "mean", "axis": ["B", "T"], "keep_dims": True, "from": ["data"],
+                      "is_output_layer": True}
+    })
+    session.run(tf.global_variables_initializer())
+    out = net.layers["reduce_mean"].output.placeholder
+    n_batch = 3
+    max_seq_len = 10
+    feed = make_feed_dict(net.extern_data.data.values(), n_batch=n_batch, n_time=max_seq_len, same_time=True)
+    v = session.run(out, feed_dict=feed)
+    input_len = feed[list(feed.keys())[0]]
+    input_data = feed[list(feed.keys())[1]]
+
+    ref = np.zeros([n_batch, n_out])
+    for batch, len in enumerate(input_len):
+      ref[batch, :] = np.mean(input_data[batch, :len, :], axis=0)
+    ref = np.expand_dims(ref, axis=1)
+    ref = ref.mean(axis=0, keepdims=True)
```

Comment:
This is not correct. This is the mean of the mean. Which is not the same as the mean over all frames directly. I wonder why you test still passes. Maybe `rtol=1e-5` is too large? Check that.

Reply:
You are right. I guess then it is easier to just keep the flattening for cases where `axes` contains the time dimension plus other dimensions. What do you think?

Regarding the tolerance, `rtol=1e-6` still passes, `1e-7` does not. I took `1e-5` from the other usage of `assert_allclose()` in this file. Not sure what a sensible threshold is. Anyway, it should not be related to calculating the mean of the mean because currently that is also what is done in the layer.

Reply:
> You are right. I guess then it is easier to just keep the flattening for cases where `axes` contains the time dimension plus other dimensions. What do you think?

Are you talking about the test, or the layer code?

I actually did not realize, the whole logic in the layer for `tf.reduce_mean` is wrong, as you do it the same way there.

Yes,  maybe the flattening is nicer and simpler. I also thought about numerical stability, which is probably better that way.

> Regarding the tolerance, `rtol=1e-6` still passes, `1e-7` does not. I took `1e-5` from the other usage of `assert_allclose()` in this file. Not sure what a sensible threshold is. Anyway, it should not be related to calculating the mean of the mean because currently that is also what is done in the layer.

I think it's ok. I thought that you do it correct in the layer, but wrong here, so I wondered that it did not catch that. But I see now that you do it just as wrong in the layer.